fuel-vm/src/interpreter/memory.rs

```diff
@@ -265,26 +315,104 @@ where
     }
 }
 
-pub(crate) fn stack_pointer_overflow<F>(
+/// Update stack pointer, checking for validity first.
+pub(crate) fn try_update_stack_pointer(
     mut sp: RegMut<SP>,
     ssp: Reg<SSP>,
     hp: Reg<HP>,
+    new_sp: Word,
+) -> Result<(), RuntimeError> {
+    if new_sp >= *hp || new_sp < *ssp {
+        Err(PanicReason::MemoryOverflow.into())
+    } else {
+        *sp = new_sp;
+        Ok(())
+    }
+}
+
+pub(crate) fn stack_pointer_overflow<F>(
+    sp: RegMut<SP>,
+    ssp: Reg<SSP>,
+    hp: Reg<HP>,
     pc: RegMut<PC>,
     f: F,
     v: Word,
 ) -> Result<(), RuntimeError>
 where
     F: FnOnce(Word, Word) -> (Word, bool),
 {
-    let (result, overflow) = f(*sp, v);
+    let (new_sp, overflow) = f(*sp, v);
 
-    if overflow || result >= *hp || result < *ssp {
-        Err(PanicReason::MemoryOverflow.into())
-    } else {
-        *sp = result;
+    if overflow {
+        return Err(PanicReason::MemoryOverflow.into())
+    }
 
-        inc_pc(pc)
+    try_update_stack_pointer(sp, ssp, hp, new_sp)?;
+    inc_pc(pc)
+}
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn push_selected_registers(
+    memory: &mut [u8; MEM_SIZE],
+    sp: RegMut<SP>,
+    ssp: Reg<SSP>,
+    hp: Reg<HP>,
+    pc: RegMut<PC>,
+    program_regs: &ProgramRegisters,
+    segment: ProgramRegistersSegment,
+    bitmask: Imm24,
+) -> Result<(), RuntimeError> {
+    let bitmask = bitmask.to_u32();
+
+    // First update the new stack pointer, as that's the only error condition
+    let count = bitmask.count_ones();
+    let stack_range = MemoryRange::new(*sp, (count as u64) * 8)?;
+    try_update_stack_pointer(sp, ssp, hp, stack_range.words().end)?;
+
+    // Write the registers to the stack
+    let mut it = memory[stack_range.usizes()].chunks_exact_mut(8);
```

Comment:
Is it possible to cross the heap here? Or is it not because we checked that inside of the `try_update_stack_pointer`?